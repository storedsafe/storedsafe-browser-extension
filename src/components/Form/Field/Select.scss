@import '../../../variables';

.field-select .field-control {
  -moz-appearance: none; // stylelint-disable-line
  -webkit-appearance: none; // stylelint-disable-line
  appearance: none;
  padding-right: $spacing * 2 + 0.5;
  width: 100%;

  border: 2px solid $border;
  border-radius: $border-radius;
  background-color: $bg;
  color: $fg;
  font-weight: 800;

  &:hover,
  &:focus,
  &:active {
    box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.1);
  }

  &:hover {
    border-color: $border-hover;
    background-color: $bg-hover;
  }

  &:focus {
    border-color: $border-focus;
    background-color: $bg-focus;
  }

  &:active {
    border-color: $border-active;
    background-color: $bg-active;
  }

  option {
    background: $bg;
    color: $fg;
  }
}

.custom-select {
  position: relative;

  &::after {
    border: solid transparent;
    border-top-color: $fg;
    border-width: 0.4em 0.25em;
    box-sizing: border-box;
    content: '';
    height: 0.8em;
    margin-top: -0.2em;
    position: absolute;
    right: 0.75em;
    top: 50%;
    transition: $transition-duration;
    width: 0.5em;
    pointer-events: none;
  }
}
