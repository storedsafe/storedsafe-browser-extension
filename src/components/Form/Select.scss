@import '../../variables';

.select {
  .custom-select select {
    -moz-appearance: none; // stylelint-disable-line
    -webkit-appearance: none; // stylelint-disable-line
    appearance: none;
    padding-right: $spacing * 2 + $arrow-width;
    width: 100%;
  }

  .custom-select {
    position: relative;
    width: 100%;
  }

  .custom-select::after {
    @extend %arrow;
    transition: $transition-duration;
  }
}
