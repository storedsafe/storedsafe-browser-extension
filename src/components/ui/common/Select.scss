@import '../../../variables';

.select {
  .custom-select select {
    -moz-appearance: none; // stylelint-disable-line
    -webkit-appearance: none; // stylelint-disable-line
    appearance: none;
    padding-right: $spacing * 2 + $arrow-width;
    width: 100%;
    border: $border-width solid $color-border;
    cursor: pointer;
    transition: border-color $transition-duration;

    &:hover {
      border-color: $color-border-hover;
    }

    &:focus {
      border-color: $color-border-focus;
    }
  }

  .custom-select {
    position: relative;
    width: 100%;
  }

  .custom-select::after {
    @extend %arrow;
    transition: $transition-duration;
  }
}
