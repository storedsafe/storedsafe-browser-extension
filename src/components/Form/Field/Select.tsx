import * as React from 'react';
import { FieldProps, FieldAttributes } from './Field';
import './Select.scss';

type SelectProps = Pick<FieldProps, 'name' | 'options' | 'value' | 'onChange' | 'onBlur' | 'label'>;
export const Select: React.SFC<SelectProps | FieldAttributes> = ({
  name,
  options,
  value,
  label,
  onChange,
  onBlur,
  ...attributes
}: SelectProps) => (
  <div className="field field-select">
    <label htmlFor={name}>
      <span className="field-label">{label}</span>
      <div className="field-container">
        <div className="custom-select">
          <select
            className="field-control"
            id={name}
            name={name}
            value={value as string | number}
            onChange={({ target }: React.ChangeEvent<HTMLSelectElement>): void => {
              onChange && onChange(target.value, name);
            }}
            onBlur={({ target }: React.FocusEvent<HTMLSelectElement>): void => {
              onBlur && onBlur(target.value, name);
            }}
            {...attributes}>
            {options.map((option) => (
              <option key={option.value} value={option.value}>
                {option.title}
              </option>
            ))}
          </select>
        </div>
      </div>
    </label>
  </div>
);

Select.defaultProps = {
  options: [],
};
