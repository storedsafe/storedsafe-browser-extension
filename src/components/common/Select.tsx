import React from 'react'
import './Select.scss'

type SelectProps = React.DetailedHTMLProps<React.SelectHTMLAttributes<HTMLSelectElement>, HTMLSelectElement> & { children: React.ReactNode }

export const Select: React.FunctionComponent<SelectProps> = ({
  children, ...props
}: SelectProps) => (
  <div className="select">
    <div className="custom-select">
      <select {...props}>
        {children}
      </select>
    </div>
  </div>
)
