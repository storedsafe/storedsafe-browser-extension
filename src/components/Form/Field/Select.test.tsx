import * as React from 'react';
import { Select } from './Select';
import { shallow, mount } from 'enzyme';

const onChange = jest.fn();
const onBlur = jest.fn();

const options = [
  { title: 'Foo', value: 'foo' },
  { title: 'Bar', value: 'bar' },
];

beforeEach(() => {
  onChange.mockClear();
  onBlur.mockClear();
});

test('<Select />', () => {
  const wrapper = shallow(<Select name="test" label="test" />);
  const select = wrapper.find('select#test');
  expect(select.find('option').length).toBe(0);
  expect(wrapper.find('div.custom-select').length).toBe(1);
  expect(wrapper).toMatchSnapshot();
});

test('<Select options=...x2 />', () => {
  const wrapper = shallow(<Select name="test" label="test" options={options} />);
  const select = wrapper.find('select#test');
  const option = select.find('option');
  expect(option.length).toBe(2);
  expect(option.at(0).props().value).toBe('foo')
  expect(option.at(0).text()).toBe('Foo')
  expect(option.at(1).props().value).toBe('bar')
  expect(option.at(1).text()).toBe('Bar')
  expect(wrapper).toMatchSnapshot();
});

test('<Select options=...x2 />, change', () => {
  const wrapper = mount(<Select name="test" label="test" options={options} value="foo" onChange={onChange} />);
  wrapper.find('select#test').simulate('change', { target: { name: 'test', value: 'bar' } });
  expect(onChange).toHaveBeenCalledWith('bar', 'test');
});

test('<Select options=...x2 />, blur', () => {
  const wrapper = mount(<Select name="test" label="test" options={options} value="foo" onBlur={onBlur} />);
  wrapper.find('select#test').simulate('blur', { target: { name: 'test', value: 'bar' } });
  expect(onBlur).toHaveBeenCalledWith('bar', 'test');
});
