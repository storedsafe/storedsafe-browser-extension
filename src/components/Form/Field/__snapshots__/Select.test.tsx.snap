// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> 1`] = `ShallowWrapper {}`;

exports[`<Select options=...x2 /> 1`] = `ShallowWrapper {}`;
